internal/api/api.go

```diff
@@ -148,19 +153,50 @@ func NewAPIWithVersion(globalConfig *conf.GlobalConfiguration, db *storage.Conne
 					if !api.config.External.AnonymousUsers.Enabled {
 						return unprocessableEntityError(ErrorCodeAnonymousProviderDisabled, "Anonymous sign-ins are disabled")
 					}
-					if _, err := api.limitHandler(limiter)(w, r); err != nil {
+					if _, err := api.limitHandler(limitAnonymousSignIns)(w, r); err != nil {
 						return err
 					}
 					return api.SignupAnonymously(w, r)
 				}
+
+				// apply ip-based rate limiting on otps
+				if _, err := api.limitHandler(limitSignups)(w, r); err != nil {
```

Comment:
One key portion of the PR happens between these lines.  Where the IP based limit handler and signup handler are applied.